Improve CORS module documentation (#19627)
OrchardCMS/OrchardCore: src/docs/reference/modules/Cors/README.md

# CORS (`OrchardCore.Cors`)

The CORS Configuration module configures [Cross-Origin Resource Sharing (CORS)](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) policies for a tenant. CORS controls whether a browser can make a cross-origin request to the tenant and expose the response to client-side code.

CORS is enforced by browsers. It is not an authentication mechanism, an authorization mechanism, or a restriction on server-to-server clients.

## Enable the module

Enable the **CORS Configuration** feature (`OrchardCore.Cors`) for each tenant that needs CORS. In the admin, go to **Configuration** > **Features**.

> After enabling the feature, go to **Settings** > **Security** > **Cross-Origin Resource Sharing**. Access to this page requires the security-critical **Managing Cors Settings** permission (`ManageCorsSettings`), which is granted to the Administrator role by default.

Reviewer: ```suggestion
After enabling the feature, go to **Settings** > **Security** > **Cross-Origin Resource Sharing**. Access to this page requires the security-critical **Managing CORS Settings** permission (`ManageCorsSettings`), which is granted to the Administrator role by default.
```

> CORS settings are stored in the site settings of the current tenant. Policies are isolated between tenants, and the module does not define an `appsettings.json` configuration section.

Reviewer: ```suggestion
CORS settings are stored in the current tenant's site settings. Policies are isolated between tenants, and the module does not define an `appsettings.json` configuration section.
```

## Create and apply a policy

Select **Add a policy**, configure the policy, and select **Save**. Saving the settings reloads the current tenant shell so that the updated policies are registered.

The module adds every configured policy to ASP.NET Core CORS by name. It applies one policy as the tenant's default:

- The policy marked **Set as default policy** is applied globally by the CORS middleware.
- If no policy is marked as the default, the first configured policy becomes the default.
- If there are no policies, the middleware does not add CORS response headers.

Define exactly one default policy when configuring multiple policies so that their application does not depend on list order.

A custom module can select another named policy for a controller or action by using ASP.NET Core endpoint metadata:

```csharp
using Microsoft.AspNetCore.Cors;

> [EnableCors("PartnerApi")]

Reviewer: Are we using them in OC? I don't think so

Reviewer: @agriffard, did you test this?

| **Allow any method** | Allows every HTTP method and ignores **Allowed methods**. |
| **Allowed methods** | Allows only the listed methods when **Allow any method** is disabled. Enter HTTP method names such as `GET`, `POST`, `PUT`, or `DELETE`. |
| **Allow any header** | Allows every request header in a CORS preflight request and ignores **Allowed headers**. |
| **Allowed headers** | Allows only the listed request headers when **Allow any header** is disabled, for example `Content-Type` or `Authorization`. |
| **Allow credentials** | Allows a browser to include credentials such as cookies or HTTP authentication information. The client must also opt in to credentialed requests, and normal cookie, authentication, and authorization rules still apply. |
| **Exposed headers** | Adds the listed response headers to `Access-Control-Expose-Headers`, which makes them readable by browser client code. CORS-safelisted response headers are readable without listing them. |

> An empty allowed-origins, allowed-methods, or allowed-headers list permits none of that category when its corresponding **Allow any** option is disabled.

Reviewer: ```suggestion
An empty allowed-origins, allowed-methods, or allowed-headers list permits none of those categories when the corresponding **Allow any** option is disabled.
```

            "AllowedOrigins": [
              "https://app.example.com"
            ],
            "AllowAnyHeader": false,
            "AllowedHeaders": [
              "Content-Type",
              "Authorization"
            ],
            "AllowAnyMethod": false,
            "AllowedMethods": [
              "GET",
              "POST"
            ],
            "AllowCredentials": true,
            "IsDefaultPolicy": true,
            "ExposedHeaders": [
              "X-Pagination-Total"
            ]
          }
        ]
      }
    }
  ]
}
```

When the **Deployment** feature is enabled, add the **Cors settings** step to a deployment plan to export these settings. The generated recipe uses the same `Settings` step and `CorsSettings` structure.

Recipe and deployment imports write site settings directly and do not use the admin form's validation. Validate imported policies, define no more than one default, and reload an already-running tenant shell after changing settings outside the admin UI.

## Troubleshooting

When a browser reports a CORS failure:

1. Confirm that the request is cross-origin. Same-origin requests do not require CORS.

> 2. Confirm that `OrchardCore.Cors` and the intended policy are configured for the tenant selected by the request's host name.

Reviewer: ```suggestion
2. Confirm that `OrchardCore.Cors` and the intended policy are configured for the tenant selected by the request's hostname.
```
